File: rust-playground/made-of-bits/src/bits.rs
use std::cmp::PartialOrd;
use std::ops::{BitAndAssign, Shr, Sub};

/// Represents a block of a BitBuf or an IntBuf. Implemented by u32 and u64.
pub(crate) trait BitBlock:
    Copy + Sized + PartialOrd<Self> + BitAndAssign + Sub<Self> + Shr<u32, Output = Self>
{
    /// Number of bits in this bit block
    const BITS: u32;
    /// Power of 2 of the number of bits in this bit block
    const BITS_LOG2: u32 = Self::BITS.ilog2();

    const ZERO: Self;
    const ONE: Self;
    const MAX: Self;

    // Delegated to the underlying block type by the individual impls
    fn saturating_sub(self, rhs: Self) -> Self;
    fn trailing_zeros(self) -> u32;

    /// Block index of the block containing the `n`-th bit
    fn block_index(n: u32) -> usize {
        // We use 'as usize' here since `n` is a u32 so this will work
        // regardless of whether usize is u32 or u64 (and we assume it
        // to be no less than u32).
        (n >> Self::BITS_LOG2) as usize
    }

    /// Bit index of the `n`-th bit within its block (masking off the high bits)
    fn block_bit_index(n: u32) -> u32;
}

macro_rules! bit_block_impl {
    ($type:ty) => {
        impl BitBlock for $type {
            const BITS: u32 = Self::BITS;
            const ZERO: Self = 0;
            const ONE: Self = 1;
            const MAX: Self = Self::MAX;

            fn saturating_sub(self, rhs: Self) -> Self {
                Self::saturating_sub(self, rhs)
            }

            fn trailing_zeros(self) -> u32 {
                Self::trailing_zeros(self)
            }

            fn block_bit_index(n: u32) -> u32 {
                n & (Self::BITS - 1)
            }
        }
    };
}

bit_block_impl!(u32);
bit_block_impl!(u64);
bit_block_impl!(u128);

/// Return the position of the k-th least significant set bit.
/// Assumes that x has at least k set Bits.
/// E.g. select1(0b1100, 0) === 2 and select1(0b1100, 1) === 3
///
/// Will panic due to overflow if the requested bit does not exist,
/// eg. select1(0b1100, 2)
///
/// As an aside, if we're interested in potentially more efficient approaches,
/// there is a broadword select1 implementation in the `succinct` package by
/// Jesse A. Tov, provided under an MIT license: https://github.com/tov/succinct-rs
///
/// An updated version of the paper is here: https://vigna.di.unimi.it/ftp/papers/Broadword.pdf
/// If we use this, here are some items for future work:
/// - Benchmark comparisons with the iterative select1 below
/// - Use simd128 to accelerate u_le8, le8, and u_nz8
/// - Implement 32-bit, 16-bit, and 8-bit select1
/// - Write my own tests (the original file had tests, but I'd like to practice writing my own)
pub(crate) fn select1<T: BitBlock>(x: T, k: u32) -> Option<u32> {
    // Unset the k-1 preceding 1-bits
    let mut x = x;
    for _ in 0..k {
        // prevent overflow when reaching for a bit that does not exist
        x &= x.saturating_sub(T::ONE);
    }
    let i = x.trailing_zeros();
    if i == T::BITS {
        // x is 0; there is no k-th bit
        None
    } else {
        Some(i)
    }
}

// todo: this doctest fails because this function is not exported!
//
// /// Reverse the lowest `num_bits` bits of `x`.
// ///
// /// ```
// /// assert_eq!(reverse_low_bits(0b0000100100, 6), 0b0000001001)
// /// //                                ^^^^^^            ^^^^^^
// /// ```
// ///
pub(crate) const fn reverse_low_bits(x: usize, num_bits: usize) -> usize {
    x.reverse_bits() >> (usize::BITS as usize - num_bits)
}

pub(crate) fn one_mask<T: BitBlock>(n: u32) -> T {
    debug_assert!(n <= T::BITS);
    if n == 0 {
        T::ZERO
    } else {
        T::MAX >> (T::BITS - n)
    }
}

/// Perform binary search over a 0..n using bitwise binary search.
/// See: https://orlp.net/blog/bitwise-binary-search/
pub(crate) fn partition_point(n: usize, pred: impl Fn(usize) -> bool) -> usize {
    let mut b = 0;
    let mut bit = bit_floor(n);
    while bit != 0 {
        let i = (b | bit) - 1;
        if i < n && pred(i) {
            b |= bit
        }
        bit >>= 1;
    }
    b
}

/// If x is not zero, calculates the largest integral power of two that is not
/// greater than x. If x is zero, returns zero.
pub(crate) const fn bit_floor(x: usize) -> usize {
    if x == 0 {
        0
    } else {
        let msb = usize::BITS - 1 - x.leading_zeros();
        1 << msb
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_one_mask() {
        for n in 0..32 {
            assert_eq!(one_mask::<u32>(n), 2u32.pow(n) - 1);
            assert_eq!(one_mask::<u32>(32), u32::MAX);
        }

        for n in 0..64 {
            assert_eq!(one_mask::<u64>(n), 2u64.pow(n) - 1);
            assert_eq!(one_mask::<u64>(64), u64::MAX);
        }
    }

    #[test]
    fn test_select1() {
        // TODO: Test other block sizes. Can use a macro to repeat tests.
        // TODO: Test for bits near the end (eg. bit 31, bit 63).

        {
            // returns None for a non-existent bit
            assert_eq!(select1::<u32>(0, 0), None);
            assert_eq!(select1::<u32>(0b11111, 5), None);
            assert_eq!(select1::<u32>(0, 0), None);
            assert_eq!(select1::<u32>(0, 0), None);
        }

        {
            // returns the index of the k-th bit (from the LSB up)
            let n = 0b0111000110010;
            assert_eq!(select1::<u32>(n, 0), Some(1));
            assert_eq!(select1::<u32>(n, 1), Some(4));
            assert_eq!(select1::<u32>(n, 2), Some(5));
            assert_eq!(select1::<u32>(n, 3), Some(9));
            assert_eq!(select1::<u32>(n, 4), Some(10));
            assert_eq!(select1::<u32>(n, 5), Some(11));
            assert_eq!(select1::<u32>(n, 6), None);
        }
    }

    #[test]
    fn test_reverse_low_bits() {
        // reverses low bits and drops high bits
        assert_eq!(
            reverse_low_bits(0b11100000000000000000000000000001, 2),
            0b0000000000000000000000000000010
        );
        assert_eq!(
            reverse_low_bits(0b11100000000000000000000000000001, 5),
            0b0000000000000000000000000010000
        );
        assert_eq!(
            reverse_low_bits(0b00000000000000000000000000000001, 3),
            0b0000000000000000000000000000100
        );
        assert_eq!(
            reverse_low_bits(0b00000000000000000000000000000101, 6),
            0b0000000000000000000000000101000
        );
    }

    #[test]
    fn test_bit_floor() {
        assert_eq!(bit_floor(0), 0);
        assert_eq!(bit_floor(1), 1);
        assert_eq!(bit_floor(2), 2);
        assert_eq!(bit_floor(3), 2);
        assert_eq!(bit_floor(4), 4);
        assert_eq!(bit_floor(5), 4);
    }

    #[test]
    fn test_partition_point() {
        let n = 100;
        let target = 60;
        assert_eq!(partition_point(n, |i| i < target), target);
        assert_eq!(partition_point(target - 1, |i| i < target), target - 1);

        assert_eq!(partition_point(0, |_| true), 0);
        assert_eq!(partition_point(1, |_| true), 1);
    }

    fn test_block_index<T: BitBlock>() {
        // zero should always be zero, regardless of block size
        assert_eq!(T::block_index(0), 0);
        // values less than a block size should map to the 0th block.
        assert_eq!(T::block_index(15), 0);
        assert_eq!(T::block_index(31), 0);
        // multiples of the block size should map to that block
        assert_eq!(T::block_index(T::BITS), 1);
        assert_eq!(T::block_index(T::BITS + 15), 1);
        assert_eq!(T::block_index(T::BITS + 31), 1);

        assert_eq!(T::block_index(2 * T::BITS), 2);
        assert_eq!(T::block_index(2 * T::BITS + 15), 2);
        assert_eq!(T::block_index(2 * T::BITS + 31), 2);
    }

    fn test_block_offset<T: BitBlock>() {
        // zero should always be zero, regardless of block size
        assert_eq!(T::block_bit_index(0), 0);
        // values less than a block size should be returned as they are.
        assert_eq!(T::block_bit_index(15), 15);
        assert_eq!(T::block_bit_index(31), 31);
        // multiples of the block size should be zero
        assert_eq!(T::block_bit_index(T::BITS), 0);
        // values above that should wrap
        assert_eq!(T::block_bit_index(T::BITS + 15), 15);
        assert_eq!(T::block_bit_index(T::BITS + 31), 31);
    }

    #[test]
    fn test_block_index_and_offset() {
        test_block_index::<u32>();
        test_block_offset::<u32>();
        test_block_index::<u64>();
        test_block_offset::<u64>();
    }
}

File: old/src/bits.js
import { assert, assertSafeInteger } from "./assert.js";

// Docs for JS bitwise operators:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#bitwise_operators

// Used by the buffer types – IntBuf and BitBuf. We don't plan to change the block type at runtime,
// so we just define the constants here, and they're imported where needed (eg. the DenseBitVec).
// It's convenient to have the block size be malleable, especially for visualization purposes, where
// being able to show examples with 8-bit blocks is a nice option.
export const BasicBlockArray = Uint32Array;
export const BasicBlockSize = 8 * BasicBlockArray.BYTES_PER_ELEMENT;
export const BasicBlockSizePow2 = Math.log2(BasicBlockSize);

/**
 * Bit index of the `n`-th bit within its block (mask off the high bits)
 * @param {number} n
 */
export function basicBlockBitOffset(n) {
  return n & (BasicBlockSize - 1);
} 

/**
 * Block index of the block containing the `n`-th bit
 * @param {number} n
 */
export function basicBlockIndex(n) {
  return n >>> BasicBlockSizePow2;
}

/**
 * Returns the largest index for which `pred` returns true, plus one.
 * If the predicate does not return true for any index, returns 0.
 * The predicate function `pred` is required to be monotonic, ie. 
 * to return `true` for all inputs below some cutoff, and `false`
 * for all inputs above that cutoff.
 * 
 * This implementation is adapted from https://orlp.net/blog/bitwise-binary-search/
 * 
 * That post contains optimized versions of this function, but here I opted for the
 * clearest implementation, at a slight performance cost.
 * 
 * @param {number} n
 * @param {(index: number) => boolean} pred
 */
export function partitionPoint(n, pred) {
  DEBUG && assert(n < 2 ** 32);
  DEBUG && assertSafeInteger(n);
  let b = 0;
  let bit = bitFloor(n);
  while (bit !== 0) {
    const i = ((b | bit) - 1) >>> 0;
    if (i < n && pred(i)) {
      b |= bit;
    }
    bit >>>= 1;
  }
  return b >>> 0;
}

/**
 * If x is not zero, calculates the largest integral power of two that is not greater than x.
 * If x is zero, returns zero.
 * Like the function in the C++ standard library: https://en.cppreference.com/w/cpp/numeric/bit_floor
 * @param {number} n
 */
export function bitFloor(n) {
  DEBUG && assert(n < 2 ** 32);
  if (n === 0) {
    return 0;
  }
  const msb = 31 - Math.clz32(n);
  return (1 << msb) >>> 0;
}

/**
 * Coerces x to an unsigned 32-bit unsigned integer. This is provided as
 * a convenience function on top of unsigned shift that does some sanity
 * checks in debug mode.
 * @param {number} x
 */
export function u32(x) {
  DEBUG && assert(Number.isInteger(x));
  // Allow bit patterns representing negative numbers, eg. 1 << 31
  DEBUG && assert(Math.abs(x) < 2 ** 32);
  return x >>> 0;
}

/**
 * Returns an unsigned 32-bit integer with its bottom `n` bits set.
 * @param {number} n
 */
export function oneMask(n) {
  DEBUG && assert(Number.isInteger(n));
  DEBUG && assert(n >= 0 && n <= 32, 'oneMask can only create masks for 32-bit integers');
  if (n === 0) return 0 >>> 0;
  return 0xffffffff >>> (32 - n);
}

/**
 * Returns the number of 1-bits in the binary representation of `x`.
 * Based on an implementation from Bit Twiddling Hacks:
 * https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel
 * An explanation of the SWAR approach: 
 * https://stackoverflow.com/questions/109023/count-the-number-of-set-bits-in-a-32-bit-integer/109025#109025
 * @param {number} n
 */
export function popcount(n) {
  DEBUG && assert(n < 2 ** 32);
  let v = (n | 0) - ((n >>> 1) & 0x55555555);
  v = (v & 0x33333333) + ((v >> 2) & 0x33333333);
  return (((v + (v >> 4)) & 0xf0f0f0f) * 0x1010101) >>> 24;
}
/**
 * Returns the number of trailing 0-bits in the binary representation of `n`.
 * Like `Math.clz32` but for trailing rather than leading zeros.
 * Based on an implementation by @mikolalysenko:
 * https://github.com/mikolalysenko/count-trailing-zeros
 * I wonder whether removing the branches would perform better:
 * eg. `c -= 16 * ((n & 0x0000ffff) !== 0); 
 * @param {number} n
 */
export function trailing0(n) {
  DEBUG && assert(n < 2 ** 32);
  n &= -n;
  let c = 32;
  if (n) c--;
  if (n & 0x0000ffff) c -= 16;
  if (n & 0x00ff00ff) c -= 8;
  if (n & 0x0f0f0f0f) c -= 4;
  if (n & 0x33333333) c -= 2;
  if (n & 0x55555555) c -= 1;
  return c;
}

// Return the position of the k-th least significant set bit.
// Assumes that x has at least k set bits.
// E.g. select1(0b1100, 0) === 2 and select1(0b1100, 1) === 3
//
// Will return 32 if the requested bit does not exist, eg. select1(0b1100, 2) === 32
//
// As an aside, if we're interested in potentially more efficient approaches,
// there is a broadword select1 implementation in the `succinct` package by
// Jesse A. Tov, provided under an MIT license: https://github.com/tov/succinct-rs
//
// An updated version of the paper is here: https://vigna.di.unimi.it/ftp/papers/Broadword.pdf
// If we use this, here are some items for future work:
// - Benchmark comparisons with the iterative select1 nelobelow
// - Use simd128 to accelerate u_le8, le8, and u_nz8
// - Implement 32-bit, 16-bit, and 8-bit select1
// - Write my own tests (the original file had tests, but I'd like to practice writing my own)

/**
 * Return the index of the `k`-th 1-bit of `n`, from the LSB upwards.
 * Returns 32 if there is no k-th 1-bit.
 * Note that the time complexity is linear in k, but constant since k <= 31.
 * https://lemire.me/blog/2018/02/21/iterating-over-set-bits-quickly/
 * @param {number} n
 * @param {number} k
 */
export function select1(n, k) {
  DEBUG && assert(n < 2 ** 32);
  // Unset the k-1 preceding 1-bits
  for (let i = 0; i < k; i++) n &= n - 1;
  return trailing0(n);
}


// Adapted from https://graphics.stanford.edu/~seander/bithacks.html#ReverseParallel
/**
 * @param {number} v
 */
export function reverseBits32(v) {
  DEBUG && assert(
    v <= 0xffffffff,
    "cannot reverse bits of an integer greater than 2^32-1"
  );
  // unsigned int v; // 32-bit word to reverse bit order
  // swap odd and even bits
  v = ((v >>> 1) & 0x55555555) | ((v & 0x55555555) << 1);
  // swap consecutive pairs
  v = ((v >>> 2) & 0x33333333) | ((v & 0x33333333) << 2);
  // swap nibbles ...
  v = ((v >>> 4) & 0x0f0f0f0f) | ((v & 0x0f0f0f0f) << 4);
  // swap bytes
  v = ((v >>> 8) & 0x00ff00ff) | ((v & 0x00ff00ff) << 8);
  // swap 2-byte long pairs
  v = (v >>> 16) | (v << 16);
  return v >>> 0;
}

/**
 * @param {number} v
 * @param {number} numBits
 */
export function reverseLowBits(v, numBits) {
  DEBUG && assert(numBits <= 32, "reverse more than 32 bits");
  return reverseBits32(v) >>> (32 - numBits);
}